### PR TITLE
Add a guessing game as an example

### PR DESCRIPTION
**Pull request by brianquinlan**
Sample game:

```
Dining, eating surface
What am I? table
Electronic, computing device
What am I? computer
Writing, drawing implement
What am I?
You missed me, I am Pencil
Learning, educational institution
What am I? school
Round, bouncy object
What am I? ball
Restful, sleep-inducing furniture
What am I? bed
Fun, plaything
What am I? toy
Shelter, living space
What am I? house
Living, breathing creature
What am I? animal
Communication, handheld device
What am I? phone
Pages of knowledge, stories
What am I? book
Light-letting, see-through opening
What am I? window
Protective, covering garments
What am I? clothes
Entrance, exit portal
What am I? door
Comfortable, sit-upon furniture
What am I? chair
Nourishing sustenance
What am I? food
Colorful, fragrant blossom
What am I? flower
Refreshing, liquid beverage
What am I? water
Nope! What am I? juice
Nope! What am I? beer
Nope! What am I? wine
Nope! What am I? coke
Nope! What am I? soda
Nope! What am I? lemonade
Nope! What am I?
You missed me, I am Drink
Tall, green, leafy
What am I? tree
Four-wheeled transport
What am I? car
Got 18 of 20 in 90 seconds
```

### CODE REVIEW

**Comment by natebosch** [comment]
It turned out to be a little awkward still to implement this with function calling, but another new feature should be exactly what we're after.
https://ai.google.dev/docs/gemini_api_overview?utm_source=gfd&utm_medium=referral&utm_campaign=blog-gemini-15-pro-preview&utm_content#json

This one should be pretty straightforward to implement compared to function calling too.

**Comment by natebosch** [comment]
> Are you going to implement it or should I give it a go?

We can probably sync on who should work on this during the weekly tomorrow. The implementation will probably want to use some of the Json Schema support that will land with function calling. I got stalled on tests but I plan on uploading a function calling PR with no/limited tests this afternoon.

**Review by natebosch — APPROVED** [review]
Would you prefer to land this as is, and I can open a separate PR to change the requests to use function calling, or do you want to roll my commits into your branch and we can merge them together?

I'm fine landing this either way - I think an implementation backed by function calling with constrained JSON decoding is a good way to implement this and we can use it as our advanced function calling example.

cc @devoncarew

File: samples/dart/bin/guessing_game.dart [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    final match = listItemExtractor.firstMatch(line);
+    if (match == null) {
+      throw FormatException('Unexpected list item: $line');
+    }
+    lists.add(match.namedGroup('content')!);
+  }
+  return lists;
+}
+
+Future<List<String>> getWords(String apiKey, String subject) async {
```
[the remark this answers is not in the document]
Comment (brianquinlan) [reply samples/dart/bin/guessing_game.dart:40]: Can't we just land the PR that you showed me before?

**Comment by brianquinlan** [comment]
> Would you prefer to land this as is, and I can open a separate PR to change the requests to use function calling, or do you want to roll my commits into your branch and we can merge them together?
> 
> I'm fine landing this either way - I think an implementation backed by function calling with constrained JSON decoding is a good way to implement this and we can use it as our advanced function calling example.
> 
> cc @devoncarew

Can't we just land the PR that you showed me before?

**Comment by natebosch** [comment]
> Can't we just land the [PR that you showed me before](https://github.com/google/generative-ai-dart/blob/guessing_game--with-tools/samples/dart/bin/guessing_game.dart)?

SGTM. Opened as a PR. https://github.com/google/generative-ai-dart/pull/125

**Pull request closed without merging**